src/http/request.cr

```diff
@@ -121,6 +117,89 @@ class HTTP::Request
     BadRequest.new
   end
 
+  private METHODS = %w(GET HEAD POST PUT DELETE CONNECT OPTIONS PATCH TRACE)
+
+  private def self.parse_request_line(io : IO) : RequestLine | BadRequest | Nil
+    # Optimization: see if we have a peek buffer
+    # (avoids a string allocation for the entire request line)
+    if peek = io.peek
+      # peek.empty? means there's no more input (EOF), so no more requests
+      return nil if peek.empty?
+
+      # See if we can find \r\n (first \n, then \r before it)
+      index = peek.index('\n'.ord.to_u8)
+      # TODO: eventually increase the 4096 bytes limit
+      if index && index > 0 && index < 4096 && peek[index - 1] == '\r'.ord.to_u8
+        parts = parse_request_line(peek[0, index - 1])
+        io.skip(index + 1) # Must skip until after \r\n
+        return parts
+      end
+    end
+
+    request_line = io.gets(4096, chomp: true)
+    return BadRequest.new unless request_line
```

Comment:
Should this be `return nil` for EOF?

Reply:
Good catch! I think this is indeed a breaking change (though it's very unlikely that the underlying `IO`s would be peekable, but still...). Fixed in #8009